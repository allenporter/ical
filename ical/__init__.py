"""
.. include:: ../README.md
"""

__all__ = [
    "alarm",
    "calendar",
    "calendar_stream",
    "event",
    "freebusy",
    "journal",
    "store",
    "timeline",
    "timespan",
    "timezone",
    "todo",
    "types",
    "tzif",
    "util",
    "diagnostics",
]
