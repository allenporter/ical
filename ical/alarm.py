"""Alarm information for calendar components."""

import datetime
import enum
from typing import Any, Optional, Union

try:
    from pydantic.v1 import Field, root_validator
except ImportError:
    from pydantic import Field, root_validator


from .component import ComponentModel
from .parsing.property import ParsedProperty
from .types import CalAddress


class Action(str, enum.Enum):
    """Type of actioninvoked when alarm is triggered."""

    AUDIO = "AUDIO"
    """An alarm that causes sound to be played to alert the user.

    The attachment is a sound resource, or a fallback is used.
    """

    DISPLAY = "DISPLAY"
    """An alarm that displays the description text to the user."""

    EMAIL = "EMAIL"
    """An email is composed and delivered to the attendees.

    The description is the body of the message, summary is the subject,
    and attachments are email attachments.
    """


class Alarm(ComponentModel):
    """An alarm component for a calendar.

    The action (e.g. AUDIO, DISPLAY, EMAIL) determine which properties
    are also specified.
    """

    action: str
    """Action to be taken when the alarm is triggered."""

    trigger: Union[datetime.timedelta, datetime.datetime]
    """May be either a relative time or absolute time."""

    duration: Optional[datetime.timedelta] = None
    """A duration in time for the alarm.

    If duration is specified then repeat must also be specified.
    """

    repeat: Optional[int] = None
    """The number of times an alarm should be repeated.

    If repeate is specified then duration must also be specified.
    """

    #
    # Properties for DISPLAY and EMAIL actions
    #

    description: Optional[str] = None
    """A description of the notification or email body."""

    #
    # Properties for EMAIL actions
    #

    summary: Optional[str] = None
    """A summary for the email action."""

    attendees: list[CalAddress] = Field(alias="attendee", default_factory=list)
    """Email recipients for the alarm."""

    extras: list[ParsedProperty] = Field(default_factory=list)

    # Future properties:
    # - attach

    @root_validator
    def parse_display_required_fields(cls, values: dict[str, Any]) -> dict[str, Any]:
        """Validate required fields for display actions."""
        action = values.get("action")
        if action != Action.DISPLAY:
            return values
        if not values.get("description"):
            raise ValueError(f"Description value is required for action {action}")
        return values

    @root_validator
    def parse_email_required_fields(cls, values: dict[str, Any]) -> dict[str, Any]:
        """Validate required fields for email actions."""
        action = values.get("action")
        if action != Action.EMAIL:
            return values
        if not values.get("description"):
            raise ValueError(f"Description value is required for action {action}")
        if not values.get("summary"):
            raise ValueError(f"Summary value is required for action {action}")
        return values

    @root_validator
    def parse_repeat_duration(cls, values: dict[str, Any]) -> dict[str, Any]:
        """Assert the relationship between repeat and duration."""
        if (values.get("duration") is None) != (values.get("repeat") is None):
            raise ValueError(
                "Duration and Repeat must both be specified or both omitted"
            )
        return values
