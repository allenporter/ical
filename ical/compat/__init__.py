"""Compatibility layer for fixing invalid ical files.

This module provides a compatibility layer for handling invalid iCalendar files.
"""

from .make_compat import enable_compat_mode

__all__ = [
    "enable_compat_mode",
]
