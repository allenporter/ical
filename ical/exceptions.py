"""Exceptions for ical library."""


class CalendarError(Exception):
    """Base exception for all ical errors."""


class CalendarParseError(CalendarError):
    """Exception raised when parsing an ical string."""


class RecurrenceError(CalendarError):
    """Exception raised when evaluating a recurrence rule.

    Recurrence rules have complex logic and it is common for there to be
    invalid date or bugs, so this special exception exists to help
    provide additional debug data to find the source of the issue. Often
    `dateutil.rrule` has limitataions and ical has to work around them
    by providing special wrapping libraries.
    """


class StoreError(CalendarError):
    """Exception thrown by a Store."""


class EventStoreError(StoreError):
    """Exception thrown by the EventStore."""


class TodoStoreError(StoreError):
    """Exception thrown by the TodoStore."""