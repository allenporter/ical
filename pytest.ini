[pytest]
log_level = DEBUG
# Can be run manually with --benchmark-only
addopts = --benchmark-skip
