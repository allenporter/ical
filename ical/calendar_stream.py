"""The core, a collection of Calendar and Scheduling objects.

This is an example of parsing an ics file as a stream of calendar objects:
```python
from pathlib import Path
from ical.calendar_stream import IcsCalendarStream

filename = Path("example/calendar.ics")
with filename.open() as ics_file:
    stream = IcsCalendarStream.from_ics(ics_file.read())
    print("File contains %s calendar(s)", len(stream.calendars))
```

You can encode a calendar stream as ics content calling the `ics()` method on
the `IcsCalendarStream`:

```python
from pathlib import Path

filename = Path("/tmp/output.ics")
with filename.open(mode="w") as ics_file:
    ics_file.write(stream.ics())
```

"""

# mypy: allow-any-generics

from __future__ import annotations

import logging

try:
    from pydantic.v1 import Field
except ImportError:
    from pydantic import Field

from .calendar import Calendar
from .component import ComponentModel
from .parsing.component import encode_content, parse_content
from .types.data_types import DATA_TYPE

_LOGGER = logging.getLogger(__name__)


class CalendarStream(ComponentModel):
    """A container that is a collection of calendaring information.

    This object supports parsing an rfc5545 calendar file, but does not
    support encoding. See `IcsCalendarStream` instead for encoding ics files.
    """

    calendars: list[Calendar] = Field(alias="vcalendar", defaut_factory=[])

    @classmethod
    def from_ics(cls, content: str) -> "CalendarStream":
        """Factory method to create a new instance from an rfc5545 iCalendar content."""
        components = parse_content(content)
        result: dict[str, list] = {"vcalendar": []}
        for component in components:
            result.setdefault(component.name, [])
            result[component.name].append(component.as_dict())
        _LOGGER.debug("Parsing object %s", result)
        return cls.parse_obj(result)

    def ics(self) -> str:
        """Encode the calendar stream as an rfc5545 iCalendar Stream content."""
        return encode_content(self.__encode_component_root__().components)


class IcsCalendarStream(CalendarStream):
    """A calendar stream that supports parsing and encoding ICS."""

    @classmethod
    def calendar_from_ics(cls, content: str) -> Calendar:
        """Load a single calendar from an ics string."""
        stream = cls.from_ics(content)
        if len(stream.calendars) == 1:
            return stream.calendars[0]
        if len(stream.calendars) == 0:
            return Calendar()
        raise ValueError("Calendar Stream had more than one calendar")

    @classmethod
    def calendar_to_ics(cls, calendar: Calendar) -> str:
        """Serialize a calendar as an ICS stream."""
        stream = cls(vcalendar=[calendar])  # type: ignore
        return stream.ics()

    class Config:
        """Configuration for IcsCalendarStream pydantic model."""

        json_encoders = DATA_TYPE.encode_property_json
        validate_assignment = True
        allow_population_by_field_name = True
