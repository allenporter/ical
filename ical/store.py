"""Library for managing the lifecycle of components in a calendar.

A store is like a manager for events within a Calendar, updating the necessary
properties such as modification times, sequence numbers, and ids. This higher
level API is a more convenient API than working with the lower level objects
directly.
"""

# pylint: disable=unnecessary-lambda

from __future__ import annotations

import datetime
import logging
from collections.abc import Callable
from typing import Any

from .calendar import Calendar
from .event import Event
from .todo import Todo
from .iter import RulesetIterable
from .timezone import Timezone
from .types import Range, Recur, RecurrenceId
from .tzif.timezoneinfo import TimezoneInfoError
from .util import dtstamp_factory

_LOGGER = logging.getLogger(__name__)


__all__ = [
    "EventStore",
    "EventStoreError",
    "TodoStore",
    "TodoStoreError",
]


class StoreError(Exception):
    """Exception thrown by a Store."""


class EventStoreError(StoreError):
    """Exception thrown by the EventStore."""


class TodoStoreError(StoreError):
    """Exception thrown by the TodoStore."""


class EventStore:
    """An event store manages the lifecycle of events on a Calendar.

    An `ical.calendar.Calendar` is a lower level object that can be directly
    manipulated to add/remove an `ical.event.Event`. That is, it does not
    handle updating timestamps, incrementing sequence numbers, or managing
    lifecycle of a recurring event during an update.


    Here is an example for setting up an `EventStore`:

    ```python
    import datetime
    from ical.calendar import Calendar
    from ical.event import Event
    from ical.store import EventStore
    from ical.types import Recur

    calendar = Calendar()
    store = EventStore(calendar)

    event = Event(
        summary="Event summary",
        start="2022-07-03",
        end="2022-07-04",
        rrule=Recur.from_rrule("FREQ=WEEKLY;COUNT=3"),
    )
    store.add(event)
    ```

    This will add events to the calendar:
    ```python3
    for event in calendar.timeline:
        print(event.summary, event.uid, event.recurrence_id, event.dtstart)
    ```
    With output like this:
    ```
    Event summary a521cf45-2c02-11ed-9e5c-066a07ffbaf5 20220703 2022-07-03
    Event summary a521cf45-2c02-11ed-9e5c-066a07ffbaf5 20220710 2022-07-10
    Event summary a521cf45-2c02-11ed-9e5c-066a07ffbaf5 20220717 2022-07-17
    ```

    You may also delete an event, or a specific instance of a recurring event:
    ```python
    # Delete a single instance of the recurring event
    store.delete(uid=event.uid, recurrence_id="20220710")
    ```

    Then viewing the store using the `print` example removes the individual
    instance in the event:
    ```
    Event summary a521cf45-2c02-11ed-9e5c-066a07ffbaf5 20220703 2022-07-03
    Event summary a521cf45-2c02-11ed-9e5c-066a07ffbaf5 20220717 2022-07-17
    ```
    """

    def __init__(
        self,
        calendar: Calendar,
        dtstamp_fn: Callable[[], datetime.datetime] = lambda: dtstamp_factory(),
    ):
        """Initialize the EventStore."""
        self._calendar = calendar
        self._dtstamp_fn = dtstamp_fn

    def _lookup_event(self, uid: str) -> Event | None:
        """Find the specified event by id."""
        for event in self._calendar.events:
            if event.uid == uid:
                return event
        return None

    def add(self, event: Event) -> Event:
        """Add the specified event to the calendar.

        This will handle assigning modification dates, sequence numbers, etc
        if those fields are unset.

        The store will ensure the `ical.calendar.Calendar` has the necessary
        `ical.timezone.Timezone` needed to fully specify the event time information
        when encoded.
        """
        update: dict[str, Any] = {}
        if not event.created:
            update["created"] = event.dtstamp
        if event.sequence is None:
            update["sequence"] = 0
        new_event = event.copy(update=update)
        _LOGGER.debug("Adding event: %s", new_event)
        self._ensure_timezone(event)
        self._calendar.events.append(new_event)
        return new_event

    def delete(
        self,
        uid: str,
        recurrence_id: str | None = None,
        recurrence_range: Range = Range.NONE,
    ) -> None:
        """Delete the event from the calendar.

        This method is used to delete an existing event. For a recurring event
        either the whole event or instances of an event may be deleted. To
        delete the complete range of a recurring event, the `uid` property
        for the event must be specified and the `recurrence_id` should not
        be specified. To delete an individual instances of the event the
        `recurrence_id` must be specified.

        When deleting individual instances, the range property may specify
        if deletion of just a specific instance, or a range of instances.
        """
        if not (store_event := self._lookup_event(uid)):
            raise EventStoreError(f"No existing event with uid: {uid}")

        if (
            recurrence_id
            and recurrence_range == Range.THIS_AND_FUTURE
            and RecurrenceId.to_value(recurrence_id) == store_event.dtstart
        ):
            # Editing the first instance and all forward is the same as editing the
            # entire series so don't bother forking a new event
            recurrence_id = None

        # Deleting all instances in the series
        if not recurrence_id:
            self._calendar.events.remove(store_event)
            return

        # Deleting one or more instances in the recurrence
        if not store_event.rrule:
            raise EventStoreError("Specified recurrence_id but event is not recurring")

        exdate = RecurrenceId.to_value(recurrence_id)
        if recurrence_range == Range.NONE:
            # A single recurrence instance is removed. Add an exclusion to
            # to the event.
            store_event.exdate.append(exdate)
            return

        # Assumes any recurrence deletion is valid, and that overwriting
        # the "until" value will not produce more instances. UNTIL is
        # inclusive so it can't include the specified exdate. FREQ=DAILY
        # is the lowest frequency supported so subtracting one day is
        # safe and works for both dates and datetimes.
        store_event.rrule.count = None
        store_event.rrule.until = exdate - datetime.timedelta(days=1)
        now = self._dtstamp_fn()
        store_event.dtstamp = now
        store_event.last_modified = now

    def edit(
        self,
        uid: str,
        event: Event,
        recurrence_id: str | None = None,
        recurrence_range: Range = Range.NONE,
    ) -> None:
        """Update the event with the specified uid.

        The specified event should be created with minimal fields, just
        including the fields that should be updated. The default fields such
        as `uid` and `dtstamp` may be used to set the uid for a new created event
        when updating a recurring event, or for any modification times.

        Example usage:
        ```python
        store.edit("event-uid-1", Event(summary="New Summary"))
        ```

        For a recurring event, either the whole event or individual instances
        of the event may be edited. To edit the complete range of a recurring
        event the `uid` property must be specified and the `recurrence_id` should
        not be specified. To edit an individual instances of the event the
        `recurrence_id` must be specified. The `recurrence_range` determines if
        just that individual instance is updated or all events following as well.

        The store will ensure the `ical.calendar.Calendar` has the necessary
        `ical.timezone.Timezone` needed to fully specify the event time information
        when encoded.
        """
        if not (store_event := self._lookup_event(uid)):
            raise EventStoreError(f"No existing event with uid: {uid}")

        if (
            recurrence_id
            and recurrence_range == Range.THIS_AND_FUTURE
            and RecurrenceId.to_value(recurrence_id) == store_event.dtstart
        ):
            # Editing the first instance and all forward is the same as editing the
            # entire series so don't bother forking a new event
            recurrence_id = None

        update = self._prepare_update(
            store_event, event, recurrence_id, recurrence_range
        )
        if recurrence_range == Range.NONE:
            # Changing the recurrence rule of a single event in the middle of the series
            # is not allowed. It is allowed to convert a single instance event to recurring.
            if event.rrule and store_event.rrule:
                if event.rrule.as_rrule_str() != store_event.rrule.as_rrule_str():
                    raise EventStoreError(
                        f"Can't update single instance with rrule (rrule={event.rrule})"
                    )
                event.rrule = None

        # Make a deep copy since deletion may update this objects recurrence rules
        new_event = store_event.copy(update=update, deep=True)
        if recurrence_id and new_event.rrule and new_event.rrule.count:
            # The recurring event count needs to skip any events that
            # come before the start of the new event. Use a RulesetIterable
            # to handle workarounds for dateutil.rrule limitations.
            dtstart: datetime.date | datetime.datetime = update["dtstart"]
            ruleset = RulesetIterable(
                store_event.dtstart,
                [new_event.rrule.as_rrule(store_event.dtstart)],
                [],
                [],
            )
            for dtvalue in ruleset:
                if dtvalue >= dtstart:
                    break
                new_event.rrule.count = new_event.rrule.count - 1

        self._ensure_timezone(event)

        # Editing a single instance of a recurring event is like deleting that instance
        # then adding a new instance on the specified date. If recurrence id is not
        # specified then the entire event is replaced.
        self.delete(uid, recurrence_id=recurrence_id, recurrence_range=recurrence_range)
        if recurrence_id:
            self.add(new_event)
        else:
            self._calendar.events.append(new_event)

    def _prepare_update(
        self,
        store_event: Event,
        event: Event,
        recurrence_id: str | None = None,
        recurrence_range: Range = Range.NONE,
    ) -> dict[str, Any]:
        """Prepare an update to an existing event."""
        partial_update = event.dict(exclude_unset=True)
        _LOGGER.debug("EV update=%s", event)
        update = {
            "created": store_event.dtstamp,
            "sequence": (store_event.sequence + 1) if store_event.sequence else 1,
            "last_modified": event.dtstamp,
            **partial_update,
            "dtstamp": event.dtstamp,
        }
        if "rrule" in update:
            update["rrule"] = Recur.parse_obj(update["rrule"])
        if recurrence_id:
            if not store_event.rrule:
                raise EventStoreError(
                    "Specified recurrence_id but event is not recurring"
                )
            # Forking a new event off the old event
            update["uid"] = event.uid
            if recurrence_range == Range.NONE:
                # The new event copied from the original is a single instance,
                # not recurrin
                update["rrule"] = None
            else:
                # Overwriting with a new recurring event
                update.update(
                    {
                        "sequence": 0,
                        "created": event.dtstamp,
                    }
                )

                # Adjust start and end time of the event
                dtstart: datetime.datetime | datetime.date = RecurrenceId.to_value(
                    recurrence_id
                )
                if event.dtstart:
                    dtstart = event.dtstart
                update["dtstart"] = dtstart
                # Event either has a duration (which should already be set) or has
                # an explicit end which needs to be realigned to new start time.
                if store_event.dtend:
                    update["dtend"] = dtstart + store_event.computed_duration
        return update

    def _ensure_timezone(self, event: Event) -> None:
        """Create a timezone object for the specified date if it does not already exist."""
        if (
            not isinstance(event.dtstart, datetime.datetime)
            or not event.dtstart.utcoffset()
            or not event.dtstart.tzinfo
        ):
            return

        # Verify this timezone does not already exist. The number of timezones
        # in a calendar is typically very small so iterate over the whole thing
        # to avoid any synchronization/cache issues.
        key = str(event.dtstart.tzinfo)
        for timezone in self._calendar.timezones:
            if timezone.tz_id == key:
                return

        new_timezone: Timezone
        try:
            new_timezone = Timezone.from_tzif(key)
        except TimezoneInfoError as err:
            raise EventStoreError(
                "No timezone information available for event: {key}"
            ) from err
        self._calendar.timezones.append(new_timezone)


class TodoStore:
    """A To-do store manages the lifecycle of to-dos on a Calendar."""

    def __init__(
        self,
        calendar: Calendar,
        dtstamp_fn: Callable[[], datetime.datetime] = lambda: dtstamp_factory(),
    ):
        """Initialize the TodoStore."""
        self._calendar = calendar
        self._dtstamp_fn = dtstamp_fn

    def _lookup_todo(self, uid: str) -> (int | None, Todo | None):
        """Find the specified todo by id returning the index."""
        for i, todo in enumerate(self._calendar.todos):
            if todo.uid == uid:
                return i, todo
        return None, None

    def add(self, todo: Todo) -> Todo:
        """Add the specified todo to the calendar."""
        update: dict[str, Any] = {}
        if not todo.created:
            update["created"] = todo.dtstamp
        if todo.sequence is None:
            update["sequence"] = 0
        new_todo = todo.copy(update=update)
        _LOGGER.debug("Adding todo: %s", new_todo)
        self._ensure_timezone(todo)
        self._calendar.todos.append(new_todo)
        return new_todo

    def delete(
        self,
        uid: str,
    ) -> None:
        """Delete the todo from the calendar."""
        store_index, store_todo = self._lookup_todo(uid)
        if not store_todo:
            raise TodoStoreError(f"No existing todo with uid: {uid}")
        self._calendar.todos.remove(store_todo)

    def edit(
        self,
        uid: str,
        todo: Todo,
    ) -> None:
        """Update the todo with the specified uid."""
        store_index, store_todo = self._lookup_todo(uid)
        if not store_todo:
            raise TodoStoreError(f"No existing todo with uid: {uid}")

        partial_update = todo.dict(exclude_unset=True)
        update = {
            "created": store_todo.dtstamp,
            "sequence": (store_todo.sequence + 1) if store_todo.sequence else 1,
            "last_modified": todo.dtstamp,
            **partial_update,
            "dtstamp": todo.dtstamp,
        }
        # Make a deep copy since deletion may update this objects recurrence rules
        new_todo = store_todo.copy(update=update, deep=True)

        self._ensure_timezone(todo)

        self._calendar.todos.pop(store_index)
        self._calendar.todos.insert(store_index, new_todo)

    def _ensure_timezone(self, todo: Todo) -> None:
        """Create a timezone object for the specified date if it does not already exist."""
        if (
            not isinstance(todo.due, datetime.datetime)
            or not todo.due.utcoffset()
            or not todo.due.tzinfo
        ):
            return

        # Verify this timezone does not already exist. The number of timezones
        # in a calendar is typically very small so iterate over the whole thing
        # to avoid any synchronization/cache issues.
        key = str(todo.due.tzinfo)
        for timezone in self._calendar.timezones:
            if timezone.tz_id == key:
                return

        new_timezone: Timezone
        try:
            new_timezone = Timezone.from_tzif(key)
        except TimezoneInfoError as err:
            raise TodoStoreError(
                "No timezone information available for todo: {key}"
            ) from err
        self._calendar.timezones.append(new_timezone)
