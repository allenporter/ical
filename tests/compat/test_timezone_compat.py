"""Tests for the extended timezone component."""

import pathlib

import pytest
from syrupy import SnapshotAssertion

from ical.exceptions import CalendarParseError
from ical.calendar_stream import CalendarStream, IcsCalendarStream
from ical.store import TodoStore
from ical.compat import timezone_compat

TESTDATA_PATH = pathlib.Path("tests/compat/testdata/")
OFFICE_365_EXTENDED_TIMEZONE = TESTDATA_PATH / "office_365_extended_timezone.ics"
OFFICE_365_INVALID_TIMEZONE = TESTDATA_PATH / "office_365_invalid_timezone.ics"


def test_extended_timezone_fail() -> None:
    """Test Office 365 extended timezone."""

    with pytest.raises(
        CalendarParseError,
        match="Expected DATE-TIME TZID value 'W. Europe Standard Time' to be valid timezone",
    ):
        IcsCalendarStream.calendar_from_ics(
            OFFICE_365_EXTENDED_TIMEZONE.read_text(encoding="utf-8")
        )


def test_extended_timezone_compat(snapshot: SnapshotAssertion) -> None:
    """Test Office 365 extended timezone with compat enabled."""
    with timezone_compat.enable_extended_timezones():
        calendar = IcsCalendarStream.calendar_from_ics(
            OFFICE_365_EXTENDED_TIMEZONE.read_text(encoding="utf-8")
        )
    assert IcsCalendarStream.calendar_to_ics(calendar) == snapshot


def test_invalid_timezone_fail() -> None:
    """Test Office 365 invalid timezone."""

    with pytest.raises(
        CalendarParseError,
        match="Expected DATE-TIME TZID value 'Customized Time Zone' to be valid timezone",
    ):
        IcsCalendarStream.calendar_from_ics(
            OFFICE_365_INVALID_TIMEZONE.read_text(encoding="utf-8")
        )


def test_invalid_timezone_compat(snapshot: SnapshotAssertion) -> None:
    """Test Office 365 invalid timezone."""

    with timezone_compat.enable_allow_invalid_timezones():
        calendar = IcsCalendarStream.calendar_from_ics(
            OFFICE_365_INVALID_TIMEZONE.read_text(encoding="utf-8")
        )
    assert IcsCalendarStream.calendar_to_ics(calendar) == snapshot
