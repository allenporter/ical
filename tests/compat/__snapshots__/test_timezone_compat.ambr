# serializer version: 1
# name: test_extended_timezone_compat
  '''
  BEGIN:VCALENDAR
  METHOD:PUBLISH
  PRODID:Microsoft Exchange Server 2010
  VERSION:2.0
  BEGIN:VEVENT
  DTSTAMP:20250417T155647Z
  UID:040000008200E00074C5B7101A82E00800000000687C546B5596DA01000000000000000
   010000000309AE93C8C3A94489F90ADBEA30C2F2B
  DTSTART;TZID=CET:20240426T140000
  DTEND;TZID=CET:20240426T150000
  SUMMARY:Uffe
  CLASS:PUBLIC
  LOCATION:
  PRIORITY:5
  SEQUENCE:0
  STATUS:CONFIRMED
  TRANSP:OPAQUE
  END:VEVENT
  END:VCALENDAR
  '''
# ---
# name: test_invalid_timezone_compat
  '''
  BEGIN:VCALENDAR
  METHOD:PUBLISH
  PRODID:Microsoft Exchange Server 2010
  VERSION:2.0
  BEGIN:VEVENT
  DTSTAMP:20250417T155647Z
  UID:040000008200E00074C5B7101A82E00800000000687C546B5596DA01000000000000000
   010000000309AE93C8C3A94489F90ADBEA30C2F2B
  DTSTART:20240426T140000
  DTEND:20240426T150000
  SUMMARY:Uffe
  CLASS:PUBLIC
  LOCATION:
  PRIORITY:5
  SEQUENCE:0
  STATUS:CONFIRMED
  TRANSP:OPAQUE
  END:VEVENT
  BEGIN:VTIMEZONE
  TZID:W. Europe Standard Time
  BEGIN:STANDARD
  DTSTART:16010101T030000
  TZOFFSETTO:+0100
  TZOFFSETFROM:+0200
  RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=10
  END:STANDARD
  BEGIN:DAYLIGHT
  DTSTART:16010101T020000
  TZOFFSETTO:+0200
  TZOFFSETFROM:+0100
  RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=3
  END:DAYLIGHT
  END:VTIMEZONE
  BEGIN:VTIMEZONE
  TZID:UTC
  BEGIN:STANDARD
  DTSTART:16010101T000000
  TZOFFSETTO:+0000
  TZOFFSETFROM:+0000
  END:STANDARD
  BEGIN:DAYLIGHT
  DTSTART:16010101T000000
  TZOFFSETTO:+0000
  TZOFFSETFROM:+0000
  END:DAYLIGHT
  END:VTIMEZONE
  END:VCALENDAR
  '''
# ---
